src/message_filter.rs

```diff
@@ -56,92 +62,60 @@ impl<Message: Hash> MessageFilter<Message> {
+        match self.count.entry(hash_code) {
+            Entry::Occupied(entry) => {
+                let &mut (ref mut c, ref mut t) = entry.into_mut();
+                *t = now;
+                *c += 1;
+                *c
+            }
+            Entry::Vacant(entry) => entry.insert((1, now)).0,
         }
     }
 
     /// Returns the number of times this message has already been inserted.
     #[cfg(test)]
     pub fn count(&self, message: &Message) -> usize {
         let hash_code = hash(message);
-        self.entries
-            .iter()
-            .find(|t| t.hash_code == hash_code)
-            .map_or(0, |t| t.count)
+        self.count.get(&hash_code).map_or(0, |&(count, _)| count)
     }
 
     /// Removes any expired messages, then returns whether `message` exists in the filter or not.
     pub fn contains(&mut self, message: &Message) -> bool {
         self.remove_expired();
-        let hash_code = hash(message);
-        self.entries.iter().any(|entry| entry.hash_code == hash_code)
+        self.count.contains_key(&hash(message))
     }
 
     /// Remove the entry for `message`, regardless of how many times it was previously inserted.
     pub fn remove(&mut self, message: &Message) {
-        let hash_code = hash(message);
-        if let Some(index) = self.entries.iter().position(|t| t.hash_code == hash_code) {
-            let _old_val = self.entries.remove(index);
-        }
+        let _old_val = self.count.remove(&hash(message));
     }
 
     /// Clears the filter, removing all the entries.
     #[cfg(feature = "use-mock-crust")]
     pub fn clear(&mut self) {
-        self.entries.clear();
+        self.count.clear();
+        self.timeout_queue.clear();
     }
 
     fn remove_expired(&mut self) {
-        let now = SystemTime::now();
-        // The entries are sorted from oldest to newest, so just split off the vector at the
-        // first unexpired entry and the returned vector is the remaining unexpired values.  If
-        // we don't find any unexpired value, just clear the vector.
-        if let Some(at) = self.entries.iter().position(|entry| entry.expiry_point > now) {
-            self.entries = self.entries.split_off(at)
-        } else {
-            self.entries.clear();
-        }
-    }
-}
-
-struct TimestampedMessage {
-    pub hash_code: u64,
-    pub expiry_point: SystemTime,
-    /// How many copies of this message have been seen before this one.
-    pub count: usize,
-}
-
-impl TimestampedMessage {
-    pub fn new(hash_code: u64, time_to_live: Duration) -> TimestampedMessage {
-        TimestampedMessage {
-            hash_code: hash_code,
-            expiry_point: SystemTime::now() + time_to_live,
-            count: 1,
+        let now = Instant::now();
+        while let Some((hash_code, time)) = self.timeout_queue.pop_front() {
+            if now.duration_since(time) <= self.time_to_live {
+                self.timeout_queue.push_front((hash_code, time));
```

Comment:
What's the time complexity of `pop_front()` followed by `push_front()`? It may be fine, but would be nice to know whether this can cause reallocations.